### PR TITLE
Update docs about cancellation to reflect latest implementation

### PR DESCRIPTION
**Pull request by qingwei91**
Hi, I wish to update the docs to reflect the current implementation, but I am not sure how much implementation details to expose?

Another question: I notice the microsite does not work out of the box locally as the microsite project uses a non-default `baseUrl`, is this something we expect?

### CODE REVIEW

File: site/src/main/mdoc/datatypes/io.md [hunk 1]
```diff
@@ -369,11 +369,15 @@ asynchronous boundaries. It can be achieved in the following way:
...
-  We should also note that `flatMap` chains are not auto-cancelable. Asynchronous
-  boundaries are important for fairness and it's not reasonable to expect interruption
-  in its' absence. With `IO`, fairness needs to be managed explicitly, the protocol being
-  easy to follow and predictable in a WYSIWYG fashion. Try to avoid very long, 
-  or never-ending loops without it.
+  We should also note that `flatMap` chains are not auto-cancelable,
+  they are cancellable only if the `flatMap` chain happens *after*
+  an asynchronous boundary mentioned above. After an asynchronous
+  boundary, cancellation checks are performed on every N `flatMap`.
+  Asynchronous boundaries are important for fairness and it's not
```
Comment (qingwei91) [inline site/src/main/mdoc/datatypes/io.md]: I am not sure if this part is still needed, AFAIK, the new implementation still only allow cancellation after async boundary, but it is not neccesary for it to happen right after async boundary

File: site/src/main/mdoc/datatypes/io.md [hunk 2]
```diff
@@ -369,11 +369,15 @@ asynchronous boundaries. It can be achieved in the following way:
   those operations is also possible to cancel. It includes, but is not limited to
   waiting on `Mvar.take`, `Mvar.put` and `Deferred.get`.
   
-  We should also note that `flatMap` chains are not auto-cancelable. Asynchronous
-  boundaries are important for fairness and it's not reasonable to expect interruption
-  in its' absence. With `IO`, fairness needs to be managed explicitly, the protocol being
-  easy to follow and predictable in a WYSIWYG fashion. Try to avoid very long, 
-  or never-ending loops without it.
+  We should also note that `flatMap` chains are not auto-cancelable,
```
Comment (Avasil) [inline site/src/main/mdoc/datatypes/io.md]: `auto-cancelable` is not really a known term or standard so I feel like we could skip this part of the sentence and just explain that `IO` checks for cancelation only per N `flatMap` and async boundaries.

File: site/src/main/mdoc/datatypes/io.md (hunk 1 above)
Comment (Avasil) [reply site/src/main/mdoc/datatypes/io.md]: Yeah, it seems to be a bit out of place. I think if we talk about fairness then we should have a small section for it somewhere

File: site/src/main/mdoc/datatypes/io.md (hunk 2 above)
Comment (qingwei91) [reply site/src/main/mdoc/datatypes/io.md]: 👍 

I've remove that sentence: https://github.com/typelevel/cats-effect/pull/730/commits/fbcdbf64849a7fd41d680c5b2ee8d55eede6c210#diff-41df762a62b991556e1f624ca666281dR372

File: site/src/main/mdoc/datatypes/io.md (hunk 1 above)
Comment (qingwei91) [reply site/src/main/mdoc/datatypes/io.md]: I removed the whole paragraph about fairness as suggested, is this ok? or would it be considered as a regression?

I am happy to create a section for it, but I am not familiar with the subject yet, do you know which part of the code should I read to have a better grasp?

If we are okay for this removal, shall I create an issue for adding a dedicated section for it?

**Comment by kubukoz** [comment]
I was under the impression it's possible to cancel a series of flatMaps without any async boundaries at all (I believe that happens in the scastie https://scastie.scala-lang.org/kGCrZ9LcQbeUp8WzlLIvjw). Also, probably an easier example than all this left/right checking would be `foreverM` or a simple recursive `val rec: IO[Unit] = IO.unit >> rec`

**Comment by qingwei91** [comment]
Hi @kubukoz 

If you are referring to the cancellation by `program.timeoutTo`, I believe it is implemented using `start` method which creates an async boundary, which makes it cancellable

**Comment by kubukoz** [comment]
@qingwei91 following that thinking, it's impossible to cancel an IO without an async boundary. So I don't think we need a distinction between flatMap chains that have one and those that don't.

File: site/src/main/mdoc/datatypes/io.md [hunk 3]
```diff
@@ -369,11 +369,28 @@ asynchronous boundaries. It can be achieved in the following way:
...
   waiting on `Mvar.take`, `Mvar.put` and `Deferred.get`.
   
-  We should also note that `flatMap` chains are not auto-cancelable. Asynchronous
-  boundaries are important for fairness and it's not reasonable to expect interruption
-  in its' absence. With `IO`, fairness needs to be managed explicitly, the protocol being
-  easy to follow and predictable in a WYSIWYG fashion. Try to avoid very long, 
-  or never-ending loops without it.
+  We should also note that `flatMap` chains are only cancelable only if
+  the chain happens *after* an asynchronous boundary.
+  After an asynchronous boundary, cancellation checks are performed on every N `flatMap`.
```
Comment (kubukoz) [inline site/src/main/mdoc/datatypes/io.md]: Worth mentioning that N is 512 at the time of writing (maxAutoCancelableBatchSize in `IORunLoop`), but could change in the future

**Comment by qingwei91** [comment]
> @qingwei91 following that thinking, it's impossible to cancel an IO without an async boundary. So I don't think we need a distinction between flatMap chains that have one and those that don't.

Hmm, from my understanding, I think there is a subtle difference, in old implementation (which is fairly old now), flatMap chain is not cancellable at all, but cancellation can happen at async boundary, for example

```scala

// Async boundary followed by `flatMap chain without Async Boundary`
IO.shift >>
flatMap(
  flatMap(
    flatMap(
      ... // 1000 more flatMap
    )
  )
)
```

If cancellation happen during the flatMap chain, ie. `after` IO.shift, then it is **NOT** cancellable with old implementation, but it is cancellable in new implementation, because the chain happens AFTER an async boundary. In old implementation, it is only cancellable if the cancellation happen before the async boundary, as that is where the check for cancellation status happen.

I realize I wasn't being clear in my previous reply

File: site/src/main/mdoc/datatypes/io.md (hunk 3 above)
Comment (qingwei91) [reply site/src/main/mdoc/datatypes/io.md]: Yep, can do, I wasn't sure how much implementation details to expose when I wrote it

**Comment by kubukoz** [comment]
Sure, that sounds correct. Old implementation aside, as far as I'm aware now we don't need any async boundaries within our IO at all to have cancelable flatMaps. It doesn't matter whether you put as shift before or after the interruption, the chain can get canceled at pretty much every step that's not within `bracket` or similar.

My point being: because we don't need an async boundary for cancelation to stop our IO at some point, I don't think this is correct:

> We should also note that `flatMap` chains are only cancelable only if the chain happens *after* an asynchronous boundary.

They are just cancelable. After an async boundary the limit is reset (I assume - didn't look for this one really) and you have N more flatMaps till the cancelation is checked, but that sentence implies you can't cancel a flatMap chain before an async boundary. 

And if we think about every IO as having an async boundary "implicitly" inserted at the beginning (if it's not started with `unsafeRunSync()`, it will have one and won't run on the main thread - it also can' be canceled if `unsafeRunSync`ed), then the sentence makes even less sense because you're always after an async boundary.

**Comment by qingwei91** [comment]
> as far as I'm aware now we don't need any async boundaries within our IO at all to have cancelable flatMaps

Hmm, I got the opposite from this PR: https://github.com/typelevel/cats-effect/pull/311, and looking into the code, it seems it is still the case

Following the call path from `unsafeRunTimed` into `step` in `IORunLoop.scala`, we can see that the cancellation check happens here https://github.com/typelevel/cats-effect/blob/master/core/shared/src/main/scala/cats/effect/internals/IORunLoop.scala#L154, inside a private method called `loop`, and this method is only invoked by `step` if there's an `Async` or `ContextSwitch`

Does it make sense? I am not really familiar with the code base, so I might have make some mistakes, I'll try to write some code to show what I meant

File: site/src/main/mdoc/datatypes/io.md (hunk 3 above)
Comment (djspiewak) [reply site/src/main/mdoc/datatypes/io.md]: In this case, I think it's definitely worth exposing the implementation detail.

File: site/src/main/mdoc/datatypes/io.md [hunk 4]
```diff
@@ -369,11 +369,28 @@ asynchronous boundaries. It can be achieved in the following way:
...
+    }
+  }
+
+  // non-terminating IO that is NOT cancelable
+  val notCancelable: IO[Int] = retryUntilRight[A, B](IO(Left(0)))
+
+  // non-terminating IO that is cancelable because there is an
+  // async boundary created by IO.shift before `flatMap` chain
+  val cancelable: IO[Int] = IO.shift *> retryUntilRight[A, B](IO(Left(0)))
+  ```
```
Comment (djspiewak) [inline site/src/main/mdoc/datatypes/io.md]: This is a really great pair of examples that nicely illustrates how subtle this can be.

**Pull request merged**